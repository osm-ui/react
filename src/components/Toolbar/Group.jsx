import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';

const Group = styled.div`
  display: flex;
  background: ${p => p.theme.toolbar.button.backgroundColor};
  box-shadow: ${p => p.theme.toolbar.boxShadow};

  &.direction-row {
    flex-direction: row;
  }
  &.direction-column {
    flex-direction: column;
  }

  &.direction-column > * {
    margin-bottom: ${p => p.theme.toolbar.childrenMargin};
  }

  &.direction-row > * {
    margin-right: ${p => p.theme.toolbar.childrenMargin};
  }

  &.direction-column > *:last-child,
  &.direction-row > *:last-child {
    margin: 0;
  }

  &.direction-column {
    &.first-shape-square {
      border-top-left-radius: ${p => p.theme.toolbar.button.borderRadius};
      border-top-right-radius: ${p => p.theme.toolbar.button.borderRadius};
    }

    &.last-shape-square {
      border-bottom-left-radius: ${p => p.theme.toolbar.button.borderRadius};
      border-bottom-right-radius: ${p => p.theme.toolbar.button.borderRadius};
    }

    &.first-shape-round {
      border-top-left-radius: ${p => p.theme.toolbar[`${p.firstSize}Size`]};
      border-top-right-radius: ${p => p.theme.toolbar[`${p.firstSize}Size`]};
    }

    &.last-shape-round {
      border-bottom-left-radius: ${p => p.theme.toolbar[`${p.lastSize}Size`]};
      border-bottom-right-radius: ${p => p.theme.toolbar[`${p.lastSize}Size`]};
    }
  }

  &.direction-row {
    &.first-shape-square {
      border-top-left-radius: ${p => p.theme.toolbar.button.borderRadius};
      border-bottom-left-radius: ${p => p.theme.toolbar.button.borderRadius};
    }

    &.last-shape-square {
      border-top-right-radius: ${p => p.theme.toolbar.button.borderRadius};
      border-bottom-right-radius: ${p => p.theme.toolbar.button.borderRadius};
    }

    &.first-shape-round {
      border-top-left-radius: ${p => p.theme.toolbar[`${p.firstSize}Size`]};
      border-bottom-left-radius: ${p => p.theme.toolbar[`${p.firstSize}Size`]};
    }

    &.last-shape-round {
      border-top-right-radius: ${p => p.theme.toolbar[`${p.lastSize}Size`]};
      border-bottom-right-radius: ${p => p.theme.toolbar[`${p.lastSize}Size`]};
    }
  }
`;

const ToolbarGroup = ({
  direction,
  size,
  shape,
  className,
  children,
  ...rest
}) => {
  let firstShape = shape;
  let lastShape = shape;
  let firstSize = size;
  let lastSize = size;

  if (!size && children.length > 0) {
    firstSize = children[0].props.size;
    lastSize = children[children.length - 1].props.size;
  }

  if (!shape && children.length > 0) {
    firstShape = children[0].props.shape;
    lastShape = children[children.length - 1].props.shape;
  }

  const classes = classnames(className, {
    [`direction-${direction}`]: true,
    [`first-shape-${firstShape}`]: true,
    [`last-shape-${lastShape}`]: true,
    [`first-size-${firstSize}`]: true,
    [`last-size-${lastSize}`]: true
  });

  const childrenProps = {
    inGroup: true
  };

  if (size) {
    childrenProps.size = size;
  }

  if (shape) {
    childrenProps.shape = shape;
  }

  return (
    <Group
      className={classes}
      firstSize={firstSize}
      lastSize={lastSize}
      {...rest}
    >
      {React.Children.map(children, child =>
        React.cloneElement(child, childrenProps)
      )}
    </Group>
  );
};

ToolbarGroup.propTypes = {
  direction: PropTypes.string,
  size: PropTypes.oneOf(['', 'xs', 'sm', 'md', 'lg']),
  shape: PropTypes.oneOf(['', 'round', 'square']),
  className: PropTypes.string,
  children: PropTypes.node.isRequired
};

ToolbarGroup.defaultProps = {
  direction: 'column',
  size: '',
  shape: '',
  className: ''
};

ToolbarGroup.displayName = 'Toolbar.Group';

export default ToolbarGroup;
