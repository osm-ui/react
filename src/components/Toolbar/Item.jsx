import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';
import FontAwesome from 'react-fontawesome';
import Loader from '../Loader';

const buildInGroupSize = props => {
  const size = props.theme.toolbar[`${props.size}Size`];
  const borderWidth = props.theme.toolbar.button.borderWidth;
  return `calc(${size} - (${borderWidth} * 2))`;
};

const StyledButton = styled.button`
  &,
  &.btn {
    position: relative;
    transition: all 0.1s ease-out;

    pointer-events: all;
    color: ${props => props.theme.toolbar.button.color};
    border-color: ${props => props.theme.toolbar.button.borderColor};
    border-style: ${props => props.theme.toolbar.button.borderStyle};
    border-width: ${props => props.theme.toolbar.button.borderWidth};
    background: ${props => props.theme.toolbar.button.backgroundColor};
    font-weight: 500;
    padding: 0;
  }

  &.shape-square {
    border-radius: ${props => props.theme.toolbar.button.borderRadius};
  }

  &.shape-round {
    border-radius: 50%;
  }

  &.xs {
    width: ${props => props.theme.toolbar.xsSize};
    height: ${props => props.theme.toolbar.xsSize};
    min-width: ${props => props.theme.toolbar.xsSize};
    min-height: ${props => props.theme.toolbar.xsSize};
    line-height: 0.85em;

    .fa {
      font-size: 0.85em;
    }
  }

  &.sm {
    width: ${props => props.theme.toolbar.smSize};
    height: ${props => props.theme.toolbar.smSize};
    min-width: ${props => props.theme.toolbar.smSize};
    min-height: ${props => props.theme.toolbar.smSize};
    line-height: 1.04em;

    .fa {
      font-size: 1.04em;
    }
  }

  &.md {
    width: ${props => props.theme.toolbar.mdSize};
    height: ${props => props.theme.toolbar.mdSize};
    min-width: ${props => props.theme.toolbar.mdSize};
    min-height: ${props => props.theme.toolbar.mdSize};
    line-height: 1.2em;

    .fa {
      font-size: 1.2em;
    }
  }

  &.lg {
    width: ${props => props.theme.toolbar.lgSize};
    height: ${props => props.theme.toolbar.lgSize};
    min-width: ${props => props.theme.toolbar.lgSize};
    min-height: ${props => props.theme.toolbar.lgSize};
    line-height: 1.43em;

    .fa {
      font-size: 1.43em;
    }
  }

  &.in-group {
    width: ${props => buildInGroupSize(props)};
    height: ${props => buildInGroupSize(props)};
    border-width: 0;
  }

  &.btn:hover {
    color: ${props => props.theme.toolbar.button.color};
    background-color: ${props =>
      props.theme.toolbar.button.hoverBackgroundColor};
    border-color: ${props => props.theme.toolbar.button.hoverBorderColor};

    &.in-group {
      border-color: ${props => props.theme.toolbar.button.hoverBackgroundColor};
    }
  }

  &.btn:focus {
    color: ${props => props.theme.toolbar.button.color};
    background-color: ${props =>
      props.theme.toolbar.button.focusBackgroundColor};
    border-color: ${props => props.theme.toolbar.button.focusBorderColor};

    &.in-group {
      border-color: ${props => props.theme.toolbar.button.focusBackgroundColor};
    }
  }

  &.btn:active,
  &.btn.active {
    color: ${props => props.theme.toolbar.button.color};
    background-color: ${props =>
      props.theme.toolbar.button.activeBackgroundColor};
    border-color: ${props => props.theme.toolbar.button.activeBorderColor};

    &.in-group {
      border-color: ${props =>
        props.theme.toolbar.button.activeBackgroundColor};
    }
  }
`;

const StyledDiv = StyledButton.extend`
  pointer-events: none;
`;

const StyledAnchor = StyledButton.extend`
  display: flex;
  align-items: center;
  justify-content: center;
  font-weight: 500;
`;

const ToolbarItem = ({
  // title,
  type,
  icon,
  size,
  shape,
  inactive,
  loading,
  // loaderTitle,
  inGroup,
  className,
  children,
  ...rest
}) => {
  const classes = classnames(className, {
    loading,
    [size]: size,
    [`shape-${shape}`]: true,
    btn: !inactive,
    'in-group': inGroup
  });

  let Element = StyledDiv;

  if (inactive === false) {
    switch (type) {
      case 'anchor':
        Element = StyledAnchor;
        break;
      default:
        Element = StyledButton;
    }
  }

  if (loading) {
    const spinnerSizes = {
      xs: 14,
      sm: 20,
      md: 24,
      lg: 30
    };
    const strokeSizes = {
      xs: 2,
      sm: 3,
      md: 3,
      lg: 3
    };
    return (
      <Element size={size} className={classes} {...rest}>
        <Loader
          spinnerSize={spinnerSizes[size]}
          strokeSize={strokeSizes[size]}
          centered
        />
      </Element>
    );
  }

  return (
    <Element size={size} className={classes} {...rest}>
      {children && children}
      {!children && <FontAwesome name={icon} />}
    </Element>
  );
};

ToolbarItem.propTypes = {
  // title: PropTypes.string,
  type: PropTypes.oneOf(['button', 'anchor']),
  icon: PropTypes.string,
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  shape: PropTypes.oneOf(['round', 'square']),
  inactive: PropTypes.bool,
  loading: PropTypes.bool,
  // loaderTitle: PropTypes.node,
  inGroup: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.node
};

ToolbarItem.defaultProps = {
  // title: '',
  type: 'button',
  icon: '',
  size: 'md',
  shape: 'round',
  inactive: false,
  loading: false,
  // loaderTitle: '',
  inGroup: false,
  className: '',
  children: ''
};

ToolbarItem.displayName = 'Toolbar.Item';

export default ToolbarItem;
