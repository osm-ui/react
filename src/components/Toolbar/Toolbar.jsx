import React from 'react';
import PropTypes from 'prop-types';
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import styled from 'styled-components';
import classnames from 'classnames';

const StyledAside = styled.aside`
  z-index: 1000;
  transition: all 0.1s ease-out;

  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  display: flex;
  margin: ${props => props.theme.toolbar.margin};
  pointer-events: none;

  &.container-parent {
    position: absolute;
  }
  &.container-root {
    position: fixed;
  }

  &.direction-row {
    flex-direction: row;

    &.position-left-top,
    &.position-left-center,
    &.position-left-bottom {
      justify-content: flex-start;
    }

    &.position-center-top,
    &.position-center-center,
    &.position-center-bottom {
      justify-content: center;
    }

    &.position-right-top,
    &.position-right-center,
    &.position-right-bottom {
      justify-content: flex-end;
    }

    &.position-left-top,
    &.position-center-top,
    &.position-right-top {
      align-items: flex-start;
    }

    &.position-left-center,
    &.position-center-center,
    &.position-right-center {
      align-items: center;
    }

    &.position-left-bottom,
    &.position-center-bottom,
    &.position-right-bottom {
      align-items: flex-end;
    }
  }

  &.direction-column {
    flex-direction: column;

    &.position-left-top,
    &.position-center-top,
    &.position-right-top {
      justify-content: flex-start;
    }

    &.position-left-center,
    &.position-center-center,
    &.position-right-center {
      justify-content: center;
    }

    &.position-left-bottom,
    &.position-center-bottom,
    &.position-right-bottom {
      justify-content: flex-end;
    }

    &.position-left-top,
    &.position-left-center,
    &.position-left-bottom {
      align-items: flex-start;
    }

    &.position-center-top,
    &.position-center-center,
    &.position-center-bottom {
      align-items: center;
    }

    &.position-right-top,
    &.position-right-center,
    &.position-right-bottom {
      align-items: flex-end;
    }
  }

  &.direction-column {
    &.position-left-top,
    &.position-left-top.transition-appear,
    &.position-left-center,
    &.position-left-center.transition-appear,
    &.position-left-bottom,
    &.position-left-bottom.transition-appear {
      transform: translate(-150%, 0);
    }

    &.position-center-top,
    &.position-center-top.transition-appear {
      transform: translate(0, -150%);
    }

    &.position-center-bottom,
    &.position-center-bottom.transition-appear {
      transform: translate(0, 150%);
    }

    &.position-right-top,
    &.position-right-top.transition-appear,
    &.position-right-center,
    &.position-right-center.transition-appear,
    &.position-right-bottom,
    &.position-right-bottom.transition-appear {
      transform: translate(150%, 0);
    }
  }

  &.direction-row {
    &.position-left-top,
    &.position-left-top.transition-appear,
    &.position-center-top,
    &.position-center-top.transition-appear,
    &.position-right-top,
    &.position-right-top.transition-appear {
      transform: translate(0, -150%);
    }

    &.position-left-center,
    &.position-left-center.transition-appear {
      transform: translate(-150%, 0);
    }

    &.position-right-center,
    &.position-right-center.transition-appear {
      transform: translate(150%, 0);
    }

    &.position-left-bottom,
    &.position-left-bottom.transition-appear,
    &.position-center-bottom,
    &.position-center-bottom.transition-appear,
    &.position-right-bottom,
    &.position-right-bottom.transition-appear {
      transform: translate(0, 150%);
    }
  }

  &.direction-column,
  &.direction-row {
    &.position-center-center,
    &.position-center-center.transition-appear {
      opacity: 0;
    }
  }

  &.direction-column,
  &.direction-row {
    &.position-left-top,
    &.position-center-top,
    &.position-right-top,
    &.position-right-center,
    &.position-right-bottom,
    &.position-center-bottom,
    &.position-left-bottom,
    &.position-left-center,
    &.position-center-center {
      &.opened,
      &.opened.transition-appear.transition-appear-active {
        opacity: ${props => props.opacity};
        transform: translate(0, 0);
      }
    }
  }

  &.direction-column > * {
    margin-bottom: ${props => props.theme.toolbar.childrenMargin};
  }

  &.direction-row > * {
    margin-right: ${props => props.theme.toolbar.childrenMargin};
  }

  &.direction-column :last-child,
  &.direction-row :last-child {
    margin: 0;
  }
`;

class Toolbar extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      opened: props.opened
    };
  }

  componentDidMount() {
    if (this.props.opened === true) {
      this._triggerCallback('onOpen');
    }
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      opened: nextProps.opened
    });

    if (this.props.opened !== nextProps.opened) {
      if (nextProps.opened === true) {
        this._triggerCallback('onOpen');
      } else {
        this._triggerCallback('onClose');
      }
    }
  }

  _triggerCallback(name) {
    if (this.props[name] !== null) {
      this.props[name]();
    }
  }

  render() {
    const {
      position,
      direction,
      size,
      shape,
      container,
      className,
      children,
      ...rest
    } = this.props;

    const asideClasses = classnames(className, {
      [`position-${position}`]: true,
      [`direction-${direction}`]: true,
      [`container-${container}`]: true,
      opened: this.state.opened
    });

    const childrenProps = {
      direction,
      position
    };

    if (size) {
      childrenProps.size = size;
    }

    if (shape) {
      childrenProps.shape = shape;
    }

    return (
      <CSSTransitionGroup
        transitionName="transition"
        transitionAppear
        transitionAppearTimeout={100}
        transitionEnter={false}
        transitionLeave={false}
      >
        <StyledAside className={asideClasses} {...rest}>
          {React.Children.map(children, child =>
            React.cloneElement(child, childrenProps)
          )}
        </StyledAside>
      </CSSTransitionGroup>
    );
  }
}

Toolbar.propTypes = {
  position: PropTypes.oneOf([
    'left-top',
    'center-top',
    'right-top',
    'right-center',
    'right-bottom',
    'center-bottom',
    'left-bottom',
    'left-center',
    'center-center'
  ]),
  direction: PropTypes.oneOf(['row', 'column']),
  size: PropTypes.oneOf(['', 'xs', 'sm', 'md', 'lg']),
  shape: PropTypes.oneOf(['', 'round', 'square']),
  container: PropTypes.oneOf(['parent', 'root']),
  opacity: PropTypes.number,
  opened: PropTypes.bool,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  className: PropTypes.string,
  children: PropTypes.node
};

Toolbar.defaultProps = {
  position: 'left-top',
  direction: 'column',
  size: '',
  shape: '',
  container: 'parent',
  opacity: 1,
  opened: false,
  onOpen: null,
  onClose: null,
  className: '',
  children: ''
};

Toolbar.displayName = 'Toolbar';

export default Toolbar;
