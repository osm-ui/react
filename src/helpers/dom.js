export function isOverflowYHidden(element) {
  return window.getComputedStyle(element).overflowY !== 'hidden';
}
