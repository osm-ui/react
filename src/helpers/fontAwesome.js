export function makePrefix(style) {
  return `fa${style[0]}`;
}

export function makeName(name) {
  return `fa-${name}`;
}
