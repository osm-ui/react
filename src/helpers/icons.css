@import url('https://use.fontawesome.com/releases/v5.2.0/css/all.css');
