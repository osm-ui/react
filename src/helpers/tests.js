import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

import DefaultTheme from 'themes/Default';

export const snapshotWithElementChildren = (TestedElement, props) => {
  const result = shallow(
    <DefaultTheme>
      <TestedElement {...props}>
        <div />
      </TestedElement>
    </DefaultTheme>
  );

  expect(result.childAt(0)).toMatchSnapshot();
};

export const snapshotWithTextChildren = (TestedElement, props) => {
  const result = shallow(
    <DefaultTheme>
      <TestedElement {...props}>This is a text</TestedElement>
    </DefaultTheme>
  );

  expect(result.childAt(0)).toMatchSnapshot();
};
