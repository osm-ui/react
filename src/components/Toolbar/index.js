import Toolbar from './Toolbar';
import ToolbarItem from './Item';
import ToolbarGroup from './Group';
import ToolbarCollapse from './Collapse';

Toolbar.Item = ToolbarItem;
Toolbar.Group = ToolbarGroup;
Toolbar.Collapse = ToolbarCollapse;

export default Toolbar;
