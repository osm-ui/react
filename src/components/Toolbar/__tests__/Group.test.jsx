import React from 'react';
// import { shallow } from 'enzyme';
import { snapshotWithElementChildren } from '../../../helpers/tests';
import Toolbar from '../';

describe('When using snapshots', () => {
  it('Should render with an element children', () =>
    snapshotWithElementChildren(Toolbar.Group));

  it('Should render with the size prop', () =>
    snapshotWithElementChildren(Toolbar.Group, { size: 'xs' }));

  it('Should render with the shape prop', () =>
    snapshotWithElementChildren(Toolbar.Group, { shape: 'square' }));

  it('Should render with the direction prop', () =>
    snapshotWithElementChildren(Toolbar.Group, { direction: 'column' }));
});
