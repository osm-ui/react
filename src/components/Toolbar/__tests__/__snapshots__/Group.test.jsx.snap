// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When using snapshots Should render with an element children 1`] = `
<Default__StyledSpan>
  <Toolbar.Group
    className=""
    direction="column"
    key=".0"
    shape=""
    size=""
  >
    <div />
  </Toolbar.Group>
</Default__StyledSpan>
`;

exports[`When using snapshots Should render with the direction prop 1`] = `
<Default__StyledSpan>
  <Toolbar.Group
    className=""
    direction="column"
    key=".0"
    shape=""
    size=""
  >
    <div />
  </Toolbar.Group>
</Default__StyledSpan>
`;

exports[`When using snapshots Should render with the shape prop 1`] = `
<Default__StyledSpan>
  <Toolbar.Group
    className=""
    direction="column"
    key=".0"
    shape="square"
    size=""
  >
    <div />
  </Toolbar.Group>
</Default__StyledSpan>
`;

exports[`When using snapshots Should render with the size prop 1`] = `
<Default__StyledSpan>
  <Toolbar.Group
    className=""
    direction="column"
    key=".0"
    shape=""
    size="xs"
  >
    <div />
  </Toolbar.Group>
</Default__StyledSpan>
`;
